src/core/xlio.h

```diff
@@ -382,6 +384,225 @@ int xlio_socketxtreme_ref_buff(struct xlio_buff_t *buff);
  */
 int xlio_socketxtreme_free_buff(struct xlio_buff_t *buff);
 
+/*
+ * Storage API
+ */
+
+typedef uintptr_t xlio_group_t;
+typedef uintptr_t xlio_socket_t;
+typedef uint32_t xlio_key_t;
+
+struct xlio_buf {
+    uint64_t userdata;
+};
+
+/*
+ * XLIO initialization.
+ *
+ * xlio_init_ex() must be called before using any storage API. This is heavy operation.
+ * xlio_init_ex() is not thread-safe operation, however, subsequent serialized calls exit
+ * successfully without any action.
+ *
+ * If set, memory_cb() notifies about memory blocks which zerocopy RX buffers can point to.
+ * Current implementation allocates a single memory block and does it within xlio_init_ex() context.
+ */
+
+/*
+ * Memory callback.
+ *
+ * XLIO calls the callback each time XLIO allocates a memory region which can be used for RX
+ * buffers. User can use this information to prepare the memory for some logic in the future.
+ * Zerocopy RX interface provides pointers to such memory.
+ *
+ * Current XLIO implementation does a single allocation for buffers.
+ */
+typedef void (*xlio_memory_cb_t)(void *addr, size_t len, size_t hugepage_size);
```

Comment:
Does xlio still support XLIO_IOCTL_USER_ALLOC? We need to allocate the memory outside of xlio, rather than just be notified of what occurred.

Reply:
At the moment it's still supported. However, I wanted to remove it eventually.

Since you need external allocation, we need to discuss the reason and possible solution, because current implementation of `XLIO_IOCTL_USER_ALLOC` is inconvenient. For example, we can simplify the external allocator (remove differentiation between TX and RX specific allocation) and provide it to `xlio_init_ex()` instead of inconvenient `ioctl()` call).